// @tensorflow/tfjs-core/dist/ops/ops.js
"use strict";
/**
 * @license
 * Copyright 2020 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
// Modularized ops.
var add_1 = require("./add");
exports.add = add_1.add;
var add_n_1 = require("./add_n");
exports.addN = add_n_1.addN;
var batchnorm_1 = require("./batchnorm");
exports.batchNorm = batchnorm_1.batchNorm;
exports.batchNormalization = batchnorm_1.batchNormalization;
var batchnorm2d_1 = require("./batchnorm2d");
exports.batchNorm2d = batchnorm2d_1.batchNorm2d;
exports.batchNormalization2d = batchnorm2d_1.batchNormalization2d;
var batchnorm3d_1 = require("./batchnorm3d");
exports.batchNorm3d = batchnorm3d_1.batchNorm3d;
exports.batchNormalization3d = batchnorm3d_1.batchNormalization3d;
var batchnorm4d_1 = require("./batchnorm4d");
exports.batchNorm4d = batchnorm4d_1.batchNorm4d;
exports.batchNormalization4d = batchnorm4d_1.batchNormalization4d;
var broadcast_to_1 = require("./broadcast_to");
exports.broadcastTo = broadcast_to_1.broadcastTo;
var clone_1 = require("./clone");
exports.clone = clone_1.clone;
var div_1 = require("./div");
exports.div = div_1.div;
var div_no_nan_1 = require("./div_no_nan");
exports.divNoNan = div_no_nan_1.divNoNan;
var eye_1 = require("./eye");
exports.eye = eye_1.eye;
var multinomial_1 = require("./multinomial");
exports.multinomial = multinomial_1.multinomial;
var one_hot_1 = require("./one_hot");
exports.oneHot = one_hot_1.oneHot;
var pad_1 = require("./pad");
exports.pad = pad_1.pad;
var pad1d_1 = require("./pad1d");
exports.pad1d = pad1d_1.pad1d;
var pad2d_1 = require("./pad2d");
exports.pad2d = pad2d_1.pad2d;
var pad3d_1 = require("./pad3d");
exports.pad3d = pad3d_1.pad3d;
var pad4d_1 = require("./pad4d");
exports.pad4d = pad4d_1.pad4d;
var rand_1 = require("./rand");
exports.rand = rand_1.rand;
var random_gamma_1 = require("./random_gamma");
exports.randomGamma = random_gamma_1.randomGamma;
var random_normal_1 = require("./random_normal");
exports.randomNormal = random_normal_1.randomNormal;
var random_uniform_1 = require("./random_uniform");
exports.randomUniform = random_uniform_1.randomUniform;
var square_1 = require("./square");
exports.square = square_1.square;
var squared_difference_1 = require("./squared_difference");
exports.squaredDifference = squared_difference_1.squaredDifference;
var tile_1 = require("./tile");
exports.tile = tile_1.tile;
var truncated_normal_1 = require("./truncated_normal");
exports.truncatedNormal = truncated_normal_1.truncatedNormal;
__export(require("./boolean_mask"));
__export(require("./complex_ops"));
__export(require("./concat_split"));
// Selectively exporting to avoid exposing gradient ops.
var conv_1 = require("./conv");
exports.conv1d = conv_1.conv1d;
exports.conv2d = conv_1.conv2d;
exports.conv3d = conv_1.conv3d;
exports.depthwiseConv2d = conv_1.depthwiseConv2d;
exports.separableConv2d = conv_1.separableConv2d;
exports.conv2dTranspose = conv_1.conv2dTranspose;
exports.conv3dTranspose = conv_1.conv3dTranspose;
__export(require("./matmul"));
__export(require("./reverse"));
__export(require("./pool"));
__export(require("./slice"));
__export(require("./unary_ops"));
__export(require("./reduction_ops"));
__export(require("./compare"));
__export(require("./binary_ops"));
__export(require("./relu_ops"));
__export(require("./logical_ops"));
__export(require("./array_ops"));
__export(require("./tensor_ops"));
__export(require("./transpose"));
__export(require("./softmax"));
__export(require("./lrn"));
__export(require("./norm"));
__export(require("./segment_ops"));
__export(require("./lstm"));
__export(require("./moving_average"));
__export(require("./strided_slice"));
__export(require("./topk"));
__export(require("./scatter_nd"));
__export(require("./spectral_ops"));
__export(require("./sparse_to_dense"));
__export(require("./gather_nd"));
__export(require("./diag"));
__export(require("./dropout"));
__export(require("./signal_ops"));
__export(require("./in_top_k"));
var operation_1 = require("./operation");
exports.op = operation_1.op;
// Second level exports.
var losses = require("./loss_ops");
exports.losses = losses;
var linalg = require("./linalg_ops");
exports.linalg = linalg;
var image = require("./image_ops");
exports.image = image;
var spectral = require("./spectral_ops");
exports.spectral = spectral;
var fused = require("./fused_ops");
exports.fused = fused;
var signal = require("./signal_ops");
exports.signal = signal;
